service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,229 @@
+	"github.com/xmidt-org/argus/chrysom"
+	"github.com/xmidt-org/argus/model"
 	"github.com/xmidt-org/webpa-common/logging"
 	"github.com/xmidt-org/webpa-common/service"
 )
 
-//datacenterWatcher checks if datacenters have been updated, based on an interval
-//If they have, then the datacenterWatcher will update the Instancers dictionary
-type datacenterWatcher struct {
+//DatacenterWatcher checks if datacenters have been updated, based on an interval
+type DatacenterWatcher struct {
+	logger                 log.Logger
+	environment            Environment
+	options                Options
+	inactiveDatacenters    map[string]bool
+	chrysomDatacenterWatch *chrysomDatacenterWatch
+	consulDatacenterWatch  *consulDatacenterWatch
+	lock                   sync.RWMutex
+}
+
+type chrysomDatacenterWatch struct {
+	chrysomClient *chrysom.Client
+	ctx           context.Context
+}
+
+type consulDatacenterWatch struct {
 	watchInterval time.Duration
-	logger        log.Logger
 	shutdown      chan struct{}
-	environment   Environment
-	options       Options
 }
 
 var (
 	defaultLogger = log.NewNopLogger()
 )
 
-func newDatacenterWatcher(logger log.Logger, environment Environment, options Options) (*datacenterWatcher, error) {
-	if options.DatacenterWatchInterval == 0 {
-		return nil, errors.New("interval cannot be 0")
-	}
+func NewDatacenterWatcher(logger log.Logger, environment Environment, options Options, ctx context.Context) (*DatacenterWatcher, error) {
+	var consulWatch *consulDatacenterWatch
 
 	if logger == nil {
 		logger = defaultLogger
 	}
 
-	return &datacenterWatcher{
-		watchInterval: options.DatacenterWatchInterval,
-		logger:        logger,
-		shutdown:      make(chan struct{}),
-		environment:   environment,
-		options:       options,
-	}, nil
+	if options.DatacenterWatchInterval > 0 {
+		consulWatch = &consulDatacenterWatch{
+			watchInterval: options.DatacenterWatchInterval,
+			shutdown:      make(chan struct{}),
+		}
+	}
+
+	datacenterWatcher := &DatacenterWatcher{
+		consulDatacenterWatch: consulWatch,
+		logger:                logger,
+		options:               options,
+		environment:           environment,
+		inactiveDatacenters:   make(map[string]bool),
+	}
+
+	if options.ChrysomConfig != nil && options.ChrysomConfig.PullInterval > 0 {
+
+		if environment.Provider() == nil {
+			return nil, errors.New("must pass in a metrics provider")
+		}
+
+		options.ChrysomConfig.MetricsProvider = environment.Provider()
+		options.ChrysomConfig.Listener = datacenterWatcher.DatacentersListener()
+		options.ChrysomConfig.Logger = logger
+		chrysomClient, err := chrysom.CreateClient(*options.ChrysomConfig)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if ctx == nil {
+			ctx = context.Background()
+		}
+
+		datacenterWatcher.chrysomDatacenterWatch = &chrysomDatacenterWatch{
+			chrysomClient: chrysomClient,
+			ctx:           ctx,
+		}
+	} else if options.ChrysomConfig != nil {
+		return nil, errors.New("chrysom pull interval cannot be 0")
```

Comment:
This checked that if a chrysom config is declared but the pull interval is 0, then it would throw an error. Hopefully with the new changes this is clearer!